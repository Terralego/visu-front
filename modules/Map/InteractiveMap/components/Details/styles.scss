.interactive-map {
  position: relative;
  overflow: hidden;

  .view-details {
    position: absolute;
    top: 0;
    bottom: 0;
    right: 0;
    width: 30rem;
    transform: translate3d(100%, 0, 0);
    transition: transform .2s ease-in;
    z-index: 99;
    padding: 1rem;
    background: #eee;

    &--visible {
      transform: translate3d(0, 0, 0);
    }

    &__close-button {
      position: absolute;
      top: 1rem;
      right: 1rem;
      background: none;
      border: none;
      cursor: pointer;
    }
  }
}
