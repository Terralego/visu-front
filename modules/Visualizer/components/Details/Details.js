import React from 'react';
import { Icon } from '@blueprintjs/core';

import MarkdownRenderer from '../MarkdownRenderer';

export const Details = ({ visible, template, onClose = () => null, ...props }) => (
  <div className={`view-details ${visible ? 'view-details--visible' : ''}`}>
    <button
      type="button"
      className="view-details__close-button"
      onClick={onClose}
    >
      <Icon icon="cross" />
    </button>
    {visible && (
      <MarkdownRenderer
        template={template}
        {...props}
      />
    )}
  </div>
);

export default Details;
