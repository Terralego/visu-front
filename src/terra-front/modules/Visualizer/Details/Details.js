import { ChevronLeft, ChevronRight, Close as CloseIcon } from '@mui/icons-material';
import { Box, Drawer, IconButton } from '@mui/material';
import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react';

import ErrorBoundary from '../../../components/ErrorBoundary';
import FeatureProperties from '../../Map/FeatureProperties';
import Template from '../../Template';

const Details = ({
  features = [],
  feature,
  interaction,
  visible,
  onClose = () => null,
  onChange = () => {},
  enableCarousel = true,
  isTableActive = false,
  translate = a => a,
}) => {
  const [index, setIndex] = useState(-1);

  // Update index when feature ID changes
  useEffect(() => {
    // eslint-disable-next-line no-underscore-dangle
    const featureId = feature && feature.properties && feature.properties._id;
    if (featureId && features.length) {
      const newIndex = features.findIndex(({ _id }) => _id === featureId);
      setIndex(newIndex);
    }
  }, [feature, features]);

  const getIndexFeature = currentIndex => {
    const featuresLength = features.length;
    if (currentIndex <= -1) {
      return featuresLength - 1;
    }
    if (currentIndex === featuresLength) {
      return 0;
    }
    return currentIndex;
  };

  const handleChange = direction => {
    const newIndex = getIndexFeature(index + direction);
    const newFeature = features[newIndex];
    if (newFeature) {
      // eslint-disable-next-line no-underscore-dangle
      onChange(newFeature._id);
      setIndex(newIndex);
    }
  };

  const isCarousel = enableCarousel && features.length > 1;
  const featureToDisplay =
    features.length > 0 && index !== -1 ? features[index] : feature && feature.properties;
  const { template, fetchProperties = {}, templates = [] } = interaction || {};

  const normalizedTemplates =
    // eslint-disable-next-line no-nested-ternary
    templates.length > 0 ? templates : template ? [{ template, fetchProperties }] : [];

  return (
    <Drawer
      variant="persistent"
      anchor="right"
      open={visible}
      sx={{
        width: 400,
        flexShrink: 0,
        '& .MuiDrawer-paper': {
          zIndex: 79,
          width: 400,
          boxSizing: 'border-box',
          backgroundColor: 'rgba(255, 255, 255, 0.9)',
          backdropFilter: 'blur(8px)',
          margin: 1,
          height: isTableActive ? '66vh' : 'calc(100vh - 16px)',
          borderRadius: 2,
          border: '1px solid rgba(0, 0, 0, 0.1)',
          boxShadow: '0 8px 32px rgba(0, 0, 0, 0.1)',
        },
      }}
    >
      <Box sx={{ display: 'flex', flexDirection: 'column', height: '100%' }}>
        <Box
          sx={{
            display: 'flex',
            alignItems: 'center',
            flexDirection: 'row',
            borderBottom: '1px solid #e0e0e0',
            margin: 1,
            marginBottom: 0,
            paddingBottom: 1,
          }}
        >
          {isCarousel && (
            <>
              <IconButton onClick={() => handleChange(-1)} size="small">
                <ChevronLeft />
              </IconButton>
              <IconButton onClick={() => handleChange(1)} size="small">
                <ChevronRight />
              </IconButton>
            </>
          )}
          <Box sx={{ flexGrow: 1 }} />
          <IconButton onClick={onClose} size="small">
            <CloseIcon />
          </IconButton>
        </Box>

        {featureToDisplay && (
          <Box
            sx={{
              display: 'flex',
              overflowX: 'hidden',
              overflowY: 'auto',
              height: '100%',
            }}
          >
            <Box
              sx={{
                padding: '8px 32px',
                flex: 'auto',
                minWidth: 0,
              }}
            >
              <Box
                sx={{
                  '& .details': {
                    display: 'flex',
                    flexDirection: 'column',
                    justifyContent: 'center',
                  },
                  '& .details__title': {
                    textAlign: 'center',
                    fontSize: '1.6rem',
                    fontWeight: 300,
                  },
                  '& .details__group': {
                    marginTop: '8px',
                  },
                  '& .details__list': {
                    padding: 0,
                    margin: 0,
                    display: 'flex',
                    flexDirection: 'column',
                  },
                  '& .details__list__amenity': {
                    display: 'flex',
                    flexDirection: 'row',
                    justifyContent: 'space-between',
                    padding: 0,
                  },
                  '& .details__list__amenity__img': {
                    width: '36px',
                    height: '36px',
                  },
                  '& .details__times': {
                    display: 'flex',
                    flexDirection: 'row',
                    justifyContent: 'space-around',
                    padding: 0,
                  },
                  '& .details__column': {
                    display: 'flex',
                    flexDirection: 'row',
                    justifyContent: 'flex-start',
                  },
                  '& .details__column-label': {
                    paddingBottom: '8px',
                    paddingRight: '8px',
                    whiteSpace: 'nowrap',
                    opacity: 0.65,
                  },
                  '& .details__column-value': {
                    minWidth: '40%',
                    paddingBottom: '10px',
                    overflowWrap: 'anywhere',
                    wordBreak: 'break-word',
                  },
                  '& .details__link': {
                    marginTop: '56px',
                    textAlign: 'center',
                    '& a': {
                      background: '#fff',
                      padding: '12px',
                      border: '1px solid rgba(10, 6, 77, 0.35)',
                    },
                  },
                  '& .details__subtitle': {
                    color: theme => theme.palette.primary.dark,
                    fontWeight: 700,
                    opacity: 0.8,
                  },
                }}
              >
                {normalizedTemplates.map((templateItem, idx) => (
                  // eslint-disable-next-line react/no-array-index-key
                  <Box key={idx} sx={{ mb: idx < normalizedTemplates.length - 1 ? 2 : 0 }}>
                    <FeatureProperties
                      {...templateItem.fetchProperties}
                      properties={featureToDisplay}
                    >
                      {({ properties: newProperties, ...staticProperties }) => (
                        <ErrorBoundary errorMsg={translate('terralego.map.template.render.error')}>
                          <Template
                            template={templateItem.template}
                            {...staticProperties}
                            {...newProperties}
                          />
                        </ErrorBoundary>
                      )}
                    </FeatureProperties>
                  </Box>
                ))}
              </Box>
            </Box>
          </Box>
        )}
      </Box>
    </Drawer>
  );
};

const propFeature = PropTypes.shape({
  properties: PropTypes.shape({
    // eslint-disable-next-line react/forbid-prop-types
    _id: PropTypes.any,
  }),
});
Details.propTypes = {
  feature: propFeature,
  interaction: PropTypes.shape({
    template: PropTypes.string,
    fetchProperties: PropTypes.shape(),
  }),
  features: PropTypes.arrayOf(propFeature),
  onChange: PropTypes.func,
  onClose: PropTypes.func,
  enableCarousel: PropTypes.bool,
  isTableActive: PropTypes.bool,
  visible: PropTypes.bool,
  translate: PropTypes.func,
};

Details.defaultProps = {
  onChange: () => {},
  onClose: () => null,
  enableCarousel: true,
  features: [],
  feature: undefined,
  interaction: undefined,
  isTableActive: false,
  visible: false,
  translate: a => a,
};

export default Details;
