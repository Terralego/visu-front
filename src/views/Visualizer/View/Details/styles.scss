.view-details {
  position: absolute;
  top: 0.5rem;
  bottom: .5rem;
  right: 0;
  width: 30rem;
  display: flex;
  flex-direction: column;
  transform: translate3d(100%, 0, 0);
  transition: transform .2s ease-in, opacity .2s ease-in;
  z-index: 99;
  background: #fff;
  margin: 0;
  opacity: 0;

  &--hidden {
    display: none;
  }

  &--visible {
    transform: translate3d(-.5rem, 0, 0);
    opacity: .9;
    box-shadow: 0 0 10px rgba(0,0,0,.2);
  }

  &__close-button {
    position: absolute;
    top: 1rem;
    right: 1rem;
    cursor: pointer;
    z-index: 1;
  }

  &__wrapper {
    display: flex;
    overflow-x: hidden;
    overflow-y: auto;
    height: 100%;
  }

  &__content {
    flex: auto;
    min-width: 0;
  }

  &__button {
    position: relative;
    transition: transform 0.2s ease-in;
    height: 100%;
    width: 30px;
    flex-shrink: 0;
    &--prev {
      transform: translateX(-100%)
    }
    &--next {
      transform: translateX(100%)
    }
    &--active {
      transform: translateX(0)
    }
    button {
      position: absolute;
      top: 50%;
      right: 0;
      left: 0;
      transform: translateY(-50%);
    }
  }

  .details {
    padding: 0.5rem;
    overflow-y: auto;
    display: flex;
    flex-direction: column;
    justify-content: center;

    &__picto {
      width: 68px;
      margin: auto;
    }

    &__loading {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%);
    }
    
    &__title {
      text-align: center;
      font-size: 1.6rem;
      font-weight: 300;
    }

    &__group {
      margin-top: .5rem;
    }

    &__list {
      padding: 0;
      margin: 0;
      display: table;
      
      &__amenity {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        padding: 0;

        &__img {
          width: 36px;
          height: 36px;
        }
      }
    }

    &__times {
      display: flex;
      flex-direction: row;
      justify-content: space-around;
      padding: 0;
    }

    &__subtitle {
      color: #106ba3;
      opacity: .65;
      font-weight: normal;
    }

    &__column {
      display: table-row;

      &-label {
        display: table-cell;
        padding-bottom: 8px;
        white-space: nowrap;
        opacity: .65;

        &:after {
          content: ' : ';
          padding-right: 1rem;
        }
      }

      &-value {
        display: table-cell;
        padding-bottom: 10px;
        overflow-wrap: anywhere; // Firefox
        word-break: break-word; // Chrome
        // If no support, it will fallback with a scrollbar

        &-time {
          display: flex;
          flex-direction: column;
          justify-content: center;
          list-style: none;
        }
      }
    }

    &__link {
      margin-top: 3.5rem;
      text-align: center;

      a {
        background: #fff;
        padding: .8rem;
        border: 1px solid rgba(10, 6, 77, .35);
      }
    }
  }
}
