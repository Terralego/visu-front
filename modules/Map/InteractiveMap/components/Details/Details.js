import React from 'react';
import { Button } from '@blueprintjs/core';

import MarkdownRenderer from '../../../../Template/MarkdownRenderer';
import './styles.scss';

export const Details = ({ visible, template, onClose = () => null, ...props }) => (
  <div className={`view-details ${visible ? 'view-details--visible' : ''}`}>
    <Button
      type="button"
      className="view-details__close-button"
      onClick={onClose}
      icon="cross"
    />
    {visible && (
      <MarkdownRenderer
        template={template}
        {...props}
      />
    )}
  </div>
);

export default Details;
