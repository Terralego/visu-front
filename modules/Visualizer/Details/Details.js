import React from 'react';
import classnames from 'classnames';
import { Button } from '@blueprintjs/core';

import FeatureProperties from '../../Map/FeatureProperties';
import Template from '../../Template';
import './styles.scss';

class Details extends React.Component {
  state = {
    index: -1,
    hidden: true,
  }

  static getDerivedStateFromProps ({ feature, interaction }) {
    /**
     * Keep a copy of props to have a nice animation with previous content
     */
    if (feature) {
      return { feature, interaction };
    }
    return null;
  }

  componentDidMount () {
    const {
      feature: { properties: { _id } = {} } = {},
    } = this.props;
    this.updateIndex(_id);
  }

  componentDidUpdate ({
    visible: prevVisible,
    feature: { properties: { _id: prevId } = {} } = {},
  }) {
    const {
      visible,
      feature: { properties: { _id } = {} } = {},
    } = this.props;

    if (prevId !== _id) {
      this.updateIndex(_id);
    }

    if (visible !== prevVisible) {
      this.switchVisibility(visible);
    }
  }

  getIndexFeature = index => {
    const { features } = this.props;
    const featuresLength = features.length;
    if (index <= -1) {
      return featuresLength - 1;
    }
    if (index === featuresLength) {
      return 0;
    }
    return index;
  }

  switchVisibility (visible) {
    if (visible) {
      clearTimeout(this.hideTimeout);
      this.setState({
        hidden: false,
      }, () => this.setState({
        visible: true,
      }));
    } else {
      this.setState({
        visible: false,
      });
      this.hideTimeout = setTimeout(() => this.setState({ hidden: true }), 200);
    }
  }

  updateIndex (id) {
    const { features = [] } = this.props;
    if (features.length) {
      this.setState({ index: features.findIndex(({ _id }) => _id === id) });
    }
  }

  handleChange (direction) {
    const { onChange } = this.props;
    const { index: prevIndex = 0 } = this.state;
    const index = this.getIndexFeature(prevIndex + direction);
    onChange(index);
    return this.setState({ index });
  }

  render () {
    const {
      features,
      onClose = () => null,
      isTableActive,
    } = this.props;
    const {
      feature: { properties } = {},
      interaction: { template, fetchProperties = {} } = {},
      index, hidden, visible,
    } = this.state;
    const isCarrousel = features.length > 1;
    const featureToDisplay = features.length > 0 && index !== -1 ? features[index] : properties;

    return (
      <div
        className={classnames(
          {
            'view-details': true,
            'bp3-light': true,
            'view-details--hidden': hidden,
            'view-details--visible': visible,
            'view-details--withTable': isTableActive,
          },
        )}
      >
        <div className="view-details__close">
          <Button
            type="button"
            className="view-details__close-button"
            onClick={onClose}
            icon="cross"
            minimal
          />
        </div>
        {featureToDisplay && (
          <div className="view-details__wrapper">
            <div
              className={classnames(
                'view-details__button',
                'view-details__button--prev',
                { 'view-details__button--active': isCarrousel },
              )}
            >
              <Button
                type="button"
                onClick={() => this.handleChange(-1)}
                icon="chevron-left"
                minimal
              />
            </div>
            <div className="view-details__content">
              <FeatureProperties
                {...fetchProperties}
                properties={featureToDisplay}
              >
                {({ properties: newProperties, ...staticProperties }) => (
                  <Template
                    template={template}
                    {...staticProperties}
                    {...newProperties}
                  />
                )}
              </FeatureProperties>
            </div>
            <div
              className={classnames(
                'view-details__button',
                'view-details__button--next',
                { 'view-details__button--active': isCarrousel },
              )}
            >
              <Button
                type="button"
                onClick={() => this.handleChange(1)}
                icon="chevron-right"
                minimal
              />
            </div>
          </div>
        )}
      </div>
    );
  }
}

export default Details;
