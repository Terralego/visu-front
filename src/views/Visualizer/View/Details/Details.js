import React from 'react';
import classnames from 'classnames';
import { Button } from '@blueprintjs/core';
import FeatureProperties from '@terralego/core/modules/Map/FeatureProperties';
import MarkdownRenderer from '@terralego/core/modules/Template/MarkdownRenderer';
import './styles.scss';

class Details extends React.Component {
  state = {
    index: -1,
    visibility: false,
  }

  componentDidMount () {
    const {
      feature: { properties: { _id } = {} } = {},
      features,
    } = this.props;
    this.setVisibility(features.some(({ _id: id }) => id === _id));
    this.updateIndex(_id);
  }

  componentDidUpdate ({
    feature: { properties: { _id: prevId } = {} } = {},
    features: prevFeatures,
  }) {
    const {
      feature: { properties: { _id } = {} } = {},
      features,
    } = this.props;
    if (prevFeatures !== features) {
      this.setVisibility(features.some(({ _id: id }) => id === _id));
    }
    if (prevId !== _id) {
      this.updateIndex(_id);
    }
  }

  getIndexFeature = index => {
    const { features } = this.props;
    const featuresLength = features.length;
    if (index <= -1) {
      return featuresLength - 1;
    }
    if (index === featuresLength) {
      return 0;
    }
    return index;
  }

  setVisibility (visibility) {
    this.setState({ visibility });
  }

  updateIndex (id) {
    const { features } = this.props;
    if (features.length) {
      this.setState({ index: features.findIndex(({ _id }) => _id === id) });
    }
  }

  handleChange (direction) {
    const { onChange } = this.props;
    const { index: prevIndex } = this.state;
    const index = this.getIndexFeature(prevIndex + direction);
    onChange(index);
    return this.setState({ index });
  }

  render () {
    const {
      visible,
      interaction: { template, fetchProperties = {} } = {},
      feature: { properties } = {},
      onClose = () => null,
      features,
    } = this.props;
    const { index, visibility } = this.state;
    const isCarrousel = features.length > 1;
    const featureToDisplay = features.length > 0 && index !== -1 ? features[index] : properties;

    const isDetailVisible = visible && (features.length === 0 || visibility);

    return (
      <div
        className={classnames(
          'view-details',
          'bp3-light',
          { 'view-details--visible': isDetailVisible },
        )}
      >
        <div className="view-details__close">
          <Button
            type="button"
            className="view-details__close-button"
            onClick={onClose}
            icon="cross"
            minimal
          />
        </div>
        {visible && (
        <div className="view-details__wrapper">
          <div
            className={classnames(
              'view-details__button',
              'view-details__button--prev',
              { 'view-details__button--active': isCarrousel },
            )}
          >
            <Button
              type="button"
              onClick={() => this.handleChange(-1)}
              icon="chevron-left"
              minimal
            />
          </div>
          <div className="view-details__content">
            <FeatureProperties
              {...fetchProperties}
              properties={featureToDisplay}
            >
              {({ properties: newProperties, ...staticProperties }) => (
                <MarkdownRenderer
                  template={template}
                  {...staticProperties}
                  {...newProperties}
                />
              )}
            </FeatureProperties>
          </div>
          <div
            className={classnames(
              'view-details__button',
              'view-details__button--next',
              { 'view-details__button--active': isCarrousel },
            )}
          >
            <Button
              type="button"
              onClick={() => this.handleChange(1)}
              icon="chevron-right"
              minimal
            />
          </div>
        </div>
        )}
      </div>
    );
  }
}

export default Details;
