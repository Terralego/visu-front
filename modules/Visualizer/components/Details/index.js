import Details from './Details';

export * from './Details';
export default Details;
