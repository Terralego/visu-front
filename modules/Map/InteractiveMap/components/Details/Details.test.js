import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Details } from './Details';

jest.mock('@blueprintjs/core', () => ({
  Button () {
    return <p>Button</p>;
  },
}));

it('should render correctly with content', () => {
  const tree = renderer.create((
    <Details
      visible
      template={
`# Hello world
 This is {{foo}}`
      }
      foo="FOO"
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render correctly without content', () => {
  const tree = renderer.create((
    <Details
      template={
`# Hello world
 This is {{foo}}`
      }
      foo="FOO"
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should close', () => {
  const onClose = jest.fn();
  const wrapper = shallow((
    <Details
      onClose={onClose}
    />
  ));
  console.log(wrapper.length);
  wrapper.find('button').props().onClick();
  expect(onClose).toHaveBeenCalled();
});

it('should close without onClose and without crashing', () => {
  const wrapper = shallow((
    <Details />
  ));
  wrapper.find('button').props().onClick();
  expect(wrapper.find('button')).toBeDefined();
});
