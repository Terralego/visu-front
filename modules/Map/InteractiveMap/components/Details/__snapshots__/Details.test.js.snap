// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with content 1`] = `
<div
  className="view-details view-details--visible"
>
  <p>
    Button
  </p>
  <h1>
    Hello world
  </h1>
  <p>
     This is FOO
  </p>
</div>
`;

exports[`should render correctly without content 1`] = `
<div
  className="view-details "
>
  <p>
    Button
  </p>
</div>
`;
